File: src/main/java/com/donato/algorithms/sorting/MergeSort.java
package com.donato.algorithms.sorting;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.stream.Collectors;

public class MergeSort {

    public static void main(String[] args) {

    }

    public static int[] mergeSortedArrays(int[] leftArray, int[] rightArray) {
        System.out.print("merge " + stringifyArr(leftArray) + " and " + stringifyArr(rightArray));
        int[] merged = new int[leftArray.length + rightArray.length];
        int leftEl = leftArray[0];
        leftArray = Arrays.copyOf(leftArray, 1);
        int rightEl = rightArray[0];
        rightArray = Arrays.copyOf(rightArray, 1);
        boolean leftEmpty = false;
        boolean rightEmpty = false;
        outer:
        while (!rightEmpty && !leftEmpty) {
            while (!rightEmpty && !leftEmpty) {
                while (rightEl <= leftEl) {
                    System.out.println(rightEl + " is smaller than or equal to " + leftEl);
                    int[] temp = merged;
                    merged = new int[merged.length + 1];
                    for (int i = 0; i < temp.length; i++) {
                        merged[i] = temp[i];
                    }
                    merged[merged.length - 1] = rightEl;
                    System.out.println("Add " + rightEl + " to merged array: [" + stringifyArr(merged) + "]");


                    if (rightArray.length > 0) {
                        rightEl = rightArray[0];
                        if (rightArray.length == 1) {
                            rightArray = new int[0];
                        } else {
                            rightArray = Arrays.copyOfRange(rightArray, 1, rightArray.length);
                        }
                    } else {
                        rightEmpty = true;
                        break outer;
                    }

                }

                while (leftEl <= rightEl) {
                    System.out.println(leftEl + " is smaller than or equal to " + rightEl);
                    int[] temp = merged;
                    merged = new int[merged.length + 1];
                    for (int i = 0; i < temp.length; i++) {
                        merged[i] = temp[i];
                    }
                    merged[merged.length - 1] = leftEl;
                    System.out.println("Add " + leftEl + " to merged array: [" + stringifyArr(merged) + "]");


                    if (leftArray.length > 0) {
                        leftEl = leftArray[0];
                        if (leftArray.length == 1) {
                            leftArray = new int[0];
                        } else {
                            leftArray = Arrays.copyOfRange(leftArray, 1, leftArray.length);
                        }
                    } else {
                        leftEmpty = true;
                        break outer;
                    }

                }


            }
        if(!leftEmpty) {
            merged = Arrays.copyOf(merged, leftArray.length + 1);
            merged[merged.length - 1] = leftEl;
        }
        if(!rightEmpty){
            Listt mergedList = new ArrayList<>(Arrays.asList(merged));
             mergedList.addAll(merged.length,Arrays.asList(rightArray))).;
        }
          return merged;
        }



    public static String stringifyArr(int[] arr){

        StringBuilder result = new StringBuilder("[");
        for (int i : arr) {
            result.append(String.valueOf(i)).append(" ");
        }
        result.append("]");
        return result.toString();
    }
}

File: src/main/java/com/donato/algorithms/sorting/MergeSort.js
function mergeSortedArrays(leftArray, rightArray) {
  console.log("merge " + leftArray + ' and ' + rightArray);
  let merged = [];
  let leftEl = leftArray.shift();
  let rightEl = rightArray.shift();

  outer: while (rightEl && leftEl) {

    while (leftEl <= rightEl) {

      console.log(leftEl + " is smaller than or equal to " + rightEl)
      merged.push(leftEl);
      console.log("add " + leftEl + " to merged array: [" + merged + "]")


      if (leftArray.length > 0)
        leftEl = leftArray.shift();
      else {
        leftEl = null;

        continue outer;
      }
    }
    while (rightEl <= leftEl) {
      console.log(rightEl + " is smaller than or equal to " + leftEl)
      merged.push(rightEl);
      console.log("add " + rightEl + " to merged array: [" + merged + "]")

      if (rightArray.length > 0)
        rightEl = rightArray.shift();
      else {
        rightEl = null;
        break outer;
      }


    }




  }

  if (leftEl) merged.push(leftEl)
  if (rightEl) merged.push(rightEl)

  if(rightArray.length > 0) {
    console.log("add "+rightArray + " to " + merged+": ")
   merged =merged.concat(rightArray)
   console.log("**"+merged+"**")
   }else if(leftArray > 0){
    console.log("add "+leftArray + " to " + merged+": ")
   merged =merged.concat(leftArray)
   console.log("**"+merged+"**")
  }
  console.log("MERGE METHOD RESULT:" + merged)

  return merged;
}

function divideArray(arr) {
  console.log("divide " + arr + " in two pieces");
  let leftHalf = arr.slice(0, arr.length / 2);
  let rightHalf = arr.slice(arr.length / 2);


  console.log("left array: " + leftHalf + ", right array: " + rightHalf)
  return [leftHalf, rightHalf];
}


function mergeSort(array) {
  console.log("call mergesort on " + array)
  if (array.length == 1) {
    console.log("BASE CASE: " + array)
    return array;
  } else {
    // Only change code below this line
    let halves = divideArray(array);
    let leftHalf = mergeSort(halves[0])

    let rightHalf = mergeSort(halves[1])
    console.log('left: ' + leftHalf);

    console.log("right: " + rightHalf);

    let mergedArray = mergeSortedArrays(leftHalf, rightHalf);

    console.log("MERGED: " + mergedArray);
    return mergedArray;
  }


}


mergeSort([1, 4, 2, 8, 345, 123, 43, 32, 5643, 63, 123, 43, 2, 55, 1, 234, 92])
